File: Java-SpringBoot/BE-02-DB/java-17-springboot-3.5/src/main/java/com/flyrank/taskservice/service/TaskServiceImpl.java
package com.flyrank.taskservice.service;

import com.flyrank.taskservice.dto.*;
import com.flyrank.taskservice.entity.TaskEntity;
import com.flyrank.taskservice.exception.InvalidTaskException;
import com.flyrank.taskservice.exception.TaskNotFoundException;
import com.flyrank.taskservice.repository.TaskRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service implementation managing task business rules and database persistence.
 */
@Service
public class TaskServiceImpl implements TaskService {

    private final TaskRepository taskRepository;

    public TaskServiceImpl(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    @Override
    @Transactional(readOnly = true)
    public List<TaskResponse> getAllTasks(String search, Boolean done) {
        return taskRepository.filterTasks(search, done).stream()
                .map(this::toResponse)
                .toList();
    }

    @Override
    @Transactional(readOnly = true)
    public TaskResponse getTaskById(Long id) {
        TaskEntity entity = taskRepository.findById(id)
                .orElseThrow(() -> new TaskNotFoundException(id));
        return toResponse(entity);
    }

    @Override
    @Transactional
    public TaskResponse createTask(TaskCreateRequest request) {
        if (request.title() == null || request.title().trim().isEmpty()) {
            throw new InvalidTaskException("Task title cannot be empty");
        }
        TaskEntity entity = new TaskEntity(request.title().trim(), false);
        TaskEntity saved = taskRepository.save(entity);
        return toResponse(saved);
    }

    @Override
    @Transactional
    public TaskResponse replaceTask(Long id, TaskReplaceRequest request) {
        TaskEntity entity = taskRepository.findById(id)
                .orElseThrow(() -> new TaskNotFoundException(id));

        if (request.title() == null || request.title().trim().isEmpty()) {
            throw new InvalidTaskException("Task title cannot be empty");
        }

        entity.setTitle(request.title().trim());
        entity.setDone(request.done());
        TaskEntity saved = taskRepository.save(entity);
        return toResponse(saved);
    }

    @Override
    @Transactional
    public TaskResponse patchTask(Long id, TaskUpdateRequest request) {
        TaskEntity entity = taskRepository.findById(id)
                .orElseThrow(() -> new TaskNotFoundException(id));

        if (request.title() != null) {
            if (request.title().trim().isEmpty()) {
                throw new InvalidTaskException("Task title cannot be empty");
            }
            entity.setTitle(request.title().trim());
        }

        if (request.done() != null) {
            entity.setDone(request.done());
        }

        TaskEntity saved = taskRepository.save(entity);
        return toResponse(saved);
    }

    @Override
    @Transactional
    public void deleteTask(Long id) {
        TaskEntity entity = taskRepository.findById(id)
                .orElseThrow(() -> new TaskNotFoundException(id));
        taskRepository.delete(entity);
    }

    @Override
    @Transactional(readOnly = true)
    public StatsResponse getStats() {
        long total = taskRepository.count();
        long done = taskRepository.countByDone(true);
        long open = total - done;
        return new StatsResponse(List.of("TASKS", "FLYWAY_SCHEMA_HISTORY"), total, done, open);
    }

    private TaskResponse toResponse(TaskEntity entity) {
        return new TaskResponse(
                entity.getId(),
                entity.getTitle(),
                entity.isDone(),
                entity.getCreatedAt(),
                entity.getUpdatedAt()
        );
    }
}

File: Java-SpringBoot/BE-02-DB/kotlin-2.3-springboot-3.5/src/main/kotlin/com/flyrank/taskservice/service/TaskServiceImpl.kt
package com.flyrank.taskservice.service

import com.flyrank.taskservice.dto.*
import com.flyrank.taskservice.entity.TaskEntity
import com.flyrank.taskservice.exception.InvalidTaskException
import com.flyrank.taskservice.exception.TaskNotFoundException
import com.flyrank.taskservice.repository.TaskRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class TaskServiceImpl(
    private val taskRepository: TaskRepository
) : TaskService {

    @Transactional(readOnly = true)
    override fun getAllTasks(search: String?, done: Boolean?): List<TaskResponse> {
        return taskRepository.filterTasks(search, done).map { toResponse(it) }
    }

    @Transactional(readOnly = true)
    override fun getTaskById(id: Long): TaskResponse {
        val entity = taskRepository.findById(id).orElseThrow { TaskNotFoundException(id) }
        return toResponse(entity)
    }

    @Transactional
    override fun createTask(request: TaskCreateRequest): TaskResponse {
        val title = request.title?.trim()
        if (title.isNullOrEmpty()) {
            throw InvalidTaskException("Task title cannot be empty")
        }
        val entity = TaskEntity(title = title, done = false)
        val saved = taskRepository.save(entity)
        return toResponse(saved)
    }

    @Transactional
    override fun replaceTask(id: Long, request: TaskReplaceRequest): TaskResponse {
        val entity = taskRepository.findById(id).orElseThrow { TaskNotFoundException(id) }

        val title = request.title?.trim()
        if (title.isNullOrEmpty()) {
            throw InvalidTaskException("Task title cannot be empty")
        }
        val done = request.done ?: throw InvalidTaskException("done status is required for PUT full replacement")

        entity.title = title
        entity.done = done
        val saved = taskRepository.save(entity)
        return toResponse(saved)
    }

    @Transactional
    override fun patchTask(id: Long, request: TaskUpdateRequest): TaskResponse {
        val entity = taskRepository.findById(id).orElseThrow { TaskNotFoundException(id) }

        request.title?.let {
            val trimmed = it.trim()
            if (trimmed.isEmpty()) {
                throw InvalidTaskException("Task title cannot be empty")
            }
            entity.title = trimmed
        }

        request.done?.let {
            entity.done = it
        }

        val saved = taskRepository.save(entity)
        return toResponse(saved)
    }

    @Transactional
    override fun deleteTask(id: Long) {
        val entity = taskRepository.findById(id).orElseThrow { TaskNotFoundException(id) }
        taskRepository.delete(entity)
    }

    @Transactional(readOnly = true)
    override fun getStats(): StatsResponse {
        val total = taskRepository.count()
        val done = taskRepository.countByDone(true)
        val open = total - done
        return StatsResponse(listOf("TASKS", "FLYWAY_SCHEMA_HISTORY"), total, done, open)
    }

    private fun toResponse(entity: TaskEntity): TaskResponse {
        return TaskResponse(
            id = requireNotNull(entity.id) { "Entity ID cannot be null" },
            title = entity.title,
            done = entity.done,
            createdAt = entity.createdAt,
            updatedAt = entity.updatedAt
        )
    }
}
